{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bring in Alignment for mapping\n",
    "\n",
    "This program will map TFBS using the Biopython's motif package.\n",
    "\n",
    "**Inputs**: \n",
    "1. before alignment (fasta) \n",
    "2. after alignment (fasta) \n",
    "3. TFBS Position Frequency Matrix.\n",
    "\n",
    "**Outputs**:\n",
    "1. `.csv` file that outputs found TFBSs at each position, if any, in alignment (position, score, species, raw_position, strand motif_found)\n",
    "2. `.csv` file that outputs only TFBS found.\n",
    "\n",
    "**To Do**:\n",
    "\n",
    "- [x] **Fix bug where species are labelled wrong only in reverse strand (how is this possible)?**\n",
    "    - I arbitrarily set the length. An artifact when I was looking at one sequence at a time.\n",
    "- [ ] **Why are there NaNs when I pivot the table? Weird.**\n",
    "- [x] Make Vector that attaches\n",
    "- [ ] Attach real species name\n",
    "- [ ] Loop through all files in directory\n",
    "- [ ] Write file name in last column\n",
    "- [ ] Append to a file \n",
    "- [ ] Have all of the important checks read out into a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import motifs\n",
    "from Bio import SeqIO \n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC, generic_dna, generic_protein\n",
    "from collections import defaultdict\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#####################\n",
    "## sys Inputs - to do\n",
    "#####################\n",
    "\n",
    "## So I can run in shell and loop through sequences AND motifs to get a giant dataset\n",
    "## Need to get the alignment to raw sequence\n",
    "\n",
    "## read in alignment as a list of sequences\n",
    "#alignment = list(SeqIO.parse(sys.argv[1], \"fasta\"))\n",
    "#motif = motifs.read(open(sys.argv[2],\"pfm\")\n",
    "\n",
    "alignment = list(SeqIO.parse(\"../data/fasta/output_ludwig_eve-striped-2.fa\", \"fasta\"))\n",
    "motif = motifs.read(open(\"../data/PWM/transpose_fm/bcd_FlyReg.fm\"),\"pfm\")\n",
    "\n",
    "raw_sequences = []\n",
    "for record in alignment:\n",
    "    raw_sequences.append(SeqRecord(record.seq.ungap(\"-\"), id = record.id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 9 records in alignment file\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n"
     ]
    }
   ],
   "source": [
    "##################\n",
    "## Input 1 - Alignment Input \n",
    "##################\n",
    "\n",
    "## Sort alphabetically\n",
    "alignment = [f for f in sorted(alignment, key=lambda x : x.id)]\n",
    "\n",
    "## Make all ids lowercase\n",
    "\n",
    "for record in alignment:\n",
    "    record.id = record.id.lower()\n",
    "\n",
    "## Check \n",
    "print(\"Found %i records in alignment file\" % len(alignment))\n",
    "\n",
    "## Sequence Length should be the same for all alignment sequences\n",
    "for seq in alignment:\n",
    "    print len(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 9 records in raw sequence file\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## Input 2 - Raw Sequences Input\n",
    "#####################\n",
    "\n",
    "print(\"Found %i records in raw sequence file\" % len(raw_sequences))\n",
    "\n",
    "## make all IUPAC.IUPACUnambiguousDNA()\n",
    "raw_sequences_2 = []\n",
    "for seq in raw_sequences:\n",
    "    raw_sequences_2.append(Seq(str(seq.seq), IUPAC.IUPACUnambiguousDNA()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        0      1      2      3      4      5      6      7\n",
      "A:   0.19   0.17   0.88   0.92   0.04   0.04   0.06   0.12\n",
      "C:   0.37   0.08   0.04   0.02   0.02   0.87   0.52   0.25\n",
      "G:   0.08   0.04   0.04   0.04   0.33   0.02   0.08   0.37\n",
      "T:   0.37   0.71   0.04   0.02   0.62   0.08   0.35   0.27\n",
      "\n",
      "        0      1      2      3      4      5      6      7\n",
      "A:  -0.38  -0.53   1.82   1.88  -2.70  -2.70  -2.12  -1.12\n",
      "C:   0.55  -1.70  -2.70  -3.70  -3.70   1.79   1.05  -0.00\n",
      "G:  -1.70  -2.70  -2.70  -2.70   0.39  -3.70  -1.70   0.55\n",
      "T:   0.55   1.51  -2.70  -3.70   1.30  -1.70   0.47   0.11\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## Input 3 - Motif Input\n",
    "#####################\n",
    "## motif.weblogo(\"mymotif.png\")\n",
    "print(motif.counts)\n",
    "pwm = motif.counts.normalize(pseudocounts=0.0) # Doesn't change from pwm\n",
    "pssm = pwm.log_odds()\n",
    "print(pssm) # Why do I need log odds exactly?\n",
    "motif_length = len(motif) #for later retrival of nucleotide sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "######################\n",
    "## Searching for Motifs in Sequences\n",
    "######################\n",
    "\n",
    "## Returns a list of arrays with a score for each position\n",
    "## This give the score for each position\n",
    "## If you print the length you get the length of the sequence minus TFBS length. \n",
    "\n",
    "## Forward stand\n",
    "pssm_list = [ ]\n",
    "for seq in raw_sequences_2:\n",
    "    pssm_list.append(pssm.calculate(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patser Threshold is 3.262\n"
     ]
    }
   ],
   "source": [
    "########################## \n",
    "## Automatic Calculation of threshold\n",
    "##########################\n",
    "\n",
    "## Ideal to find something that automatically calculates, as\n",
    "## opposed to having human choosing.\n",
    "\n",
    "## Approximate calculation of appropriate thresholds for motif finding \n",
    "## Patser Threshold\n",
    "## It selects such a threshold that the log(fpr)=-ic(M) \n",
    "## note: the actual patser software uses natural logarithms instead of log_2, so the numbers \n",
    "## are not directly comparable. \n",
    "\n",
    "distribution = pssm.distribution(background=motif.background, precision=10**4)\n",
    "patser_threshold = distribution.threshold_patser() #for use later\n",
    "\n",
    "print(\"Patser Threshold is %5.3f\" % patser_threshold) # Calculates Paster threshold. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###################################\n",
    "## Searching for motif in all raw_sequences\n",
    "#################################\n",
    "\n",
    "raw_id = []\n",
    "for seq in raw_sequences:\n",
    "    raw_id.append(seq.id)\n",
    "\n",
    "record_length = []\n",
    "for record in raw_sequences_2:\n",
    "    record_length.append(len(record))\n",
    "\n",
    "position_list = []\n",
    "for i in range(0,8):\n",
    "    for position, score in pssm.search(raw_sequences_2[i], threshold = patser_threshold):\n",
    "        positions = {'species': raw_id[i], 'score':score, 'position':position, 'seq_len': record_length[i] }\n",
    "        position_list.append(positions)\n",
    "        \n",
    "position_DF = pd.DataFrame(position_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     position      score  seq_len                           species  \\\n",
      "0          10   5.013668      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "1        -804   3.414712      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "2         145  10.457056      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "3        -740   8.946094      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "4        -664   3.572565      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "5        -598   3.594098      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "6        -580   5.417715      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "7         389   3.824103      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "8         420   4.672100      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "9        -479   4.374594      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "10        430   3.794091      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "11       -448   9.909568      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "12        460   4.094957      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "13       -436   5.787577      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "14        502   3.793224      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "15        523   3.774115      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "16       -357   9.909568      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "17        551   4.094957      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "18       -313   3.769742      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "19       -305   5.431520      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "20       -275   3.832565      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "21       -274   5.505177      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "22        635   5.016483      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "23        641  10.016483      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "24        649   3.357991      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "25       -220   3.824103      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "26       -138   5.920559      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "27        -96   3.995421      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "28        813   3.391992      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "29        -82   6.380240      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "..        ...        ...      ...                               ...   \n",
      "205       847   3.285077      898  ludwig_eve-striped-2||MEMB003D|-   \n",
      "206       888   3.875044      898  ludwig_eve-striped-2||MEMB003D|-   \n",
      "207        10   5.454241      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "208       174  10.457056      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "209      -719   8.946094      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "210      -652   5.243600      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "211      -585   3.594098      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "212      -567   6.417715      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "213       350   3.702168      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "214       444   3.794091      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "215      -439   9.909568      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "216       477   4.094957      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "217      -427   5.787577      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "218       518   5.020957      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "219      -376   4.736640      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "220      -329   9.909568      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "221       587   4.094957      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "222       597   3.346478      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "223      -294   4.374594      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "224      -262   4.389700      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "225       655   5.016483      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "226       661  10.016483      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "227      -241   8.959556      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "228       676   8.997031      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "229       760   3.559102      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "230      -115   5.920559      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "231       844   3.391992      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "232       -59   6.380240      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "233       858   5.612093      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "234       901   4.422532      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "\n",
      "     raw_position    strand  \n",
      "0              10  positive  \n",
      "1             101  negative  \n",
      "2             145  positive  \n",
      "3             165  negative  \n",
      "4             241  negative  \n",
      "5             307  negative  \n",
      "6             325  negative  \n",
      "7             389  positive  \n",
      "8             420  positive  \n",
      "9             426  negative  \n",
      "10            430  positive  \n",
      "11            457  negative  \n",
      "12            460  positive  \n",
      "13            469  negative  \n",
      "14            502  positive  \n",
      "15            523  positive  \n",
      "16            548  negative  \n",
      "17            551  positive  \n",
      "18            592  negative  \n",
      "19            600  negative  \n",
      "20            630  negative  \n",
      "21            631  negative  \n",
      "22            635  positive  \n",
      "23            641  positive  \n",
      "24            649  positive  \n",
      "25            685  negative  \n",
      "26            767  negative  \n",
      "27            809  negative  \n",
      "28            813  positive  \n",
      "29            823  negative  \n",
      "..            ...       ...  \n",
      "205           847  positive  \n",
      "206           888  positive  \n",
      "207            10  positive  \n",
      "208           174  positive  \n",
      "209           194  negative  \n",
      "210           261  negative  \n",
      "211           328  negative  \n",
      "212           346  negative  \n",
      "213           350  positive  \n",
      "214           444  positive  \n",
      "215           474  negative  \n",
      "216           477  positive  \n",
      "217           486  negative  \n",
      "218           518  positive  \n",
      "219           537  negative  \n",
      "220           584  negative  \n",
      "221           587  positive  \n",
      "222           597  positive  \n",
      "223           619  negative  \n",
      "224           651  negative  \n",
      "225           655  positive  \n",
      "226           661  positive  \n",
      "227           672  negative  \n",
      "228           676  positive  \n",
      "229           760  positive  \n",
      "230           798  negative  \n",
      "231           844  positive  \n",
      "232           854  negative  \n",
      "233           858  positive  \n",
      "234           901  positive  \n",
      "\n",
      "[235 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "#############################\n",
    "## Add strand and pos position information as columns to position_DF\n",
    "#############################\n",
    "position_list_pos = []\n",
    "for i, x in enumerate(position_DF['position']):\n",
    "    if x < 0:\n",
    "       position_list_pos.append(position_DF.loc[position_DF.index[i], 'seq_len'] + x)\n",
    "    else:\n",
    "       position_list_pos.append(x)\n",
    "\n",
    "## append to position_DF\n",
    "position_DF['raw_position'] = position_list_pos\n",
    "\n",
    "## print position_DF['raw_position']\n",
    "    \n",
    "## strand Column\n",
    "strand = []\n",
    "for x in position_DF['position']:\n",
    "    if x < 0:\n",
    "       strand.append(\"negative\")\n",
    "    else:\n",
    "       strand.append(\"positive\")\n",
    "    \n",
    "## append to position_DF\n",
    "position_DF['strand'] = strand\n",
    "\n",
    "## motif_found column\n",
    "## First turn into a list of strings\n",
    "raw_sequences_2_list = []\n",
    "for seq in raw_sequences_2:\n",
    "    raw_sequences_2_list.append(str(seq))\n",
    "\n",
    "## Now get all motifs found in sequences\n",
    "# motif_found = []\n",
    "# for x in position_DF['position']:\n",
    "    # motif_found.append(raw_sequences_2_list[i][x:x + motif_length])\n",
    "\n",
    "## append to position_DF\n",
    "# position_DF['motif_found'] = motif_found\n",
    "\n",
    "print position_DF\n",
    "\n",
    "## Check\n",
    "## len(motif_found)    \n",
    "## print(motif_found) \n",
    "## print(position_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##################\n",
    "## get alignment position \n",
    "#################\n",
    "\n",
    "## Need to map to the sequence alignment position\n",
    "\n",
    "remap_list = []\n",
    "nuc_list = ['A', 'a', 'G', 'g', 'C', 'c', 'T', 't', 'N', 'n']\n",
    "\n",
    "\n",
    "positions = {'score':score, 'position':position, 'species': i}\n",
    "position_list.append(positions)\n",
    "\n",
    "for i in range(0,9):\n",
    "    counter = 0\n",
    "    for xInd, x in enumerate(alignment[i].seq):    \n",
    "        if x in nuc_list:\n",
    "            remaps = {'raw_position': counter, 'align_position':xInd, 'species':alignment[i].id}\n",
    "            counter += 1\n",
    "            remap_list.append(remaps)\n",
    "            \n",
    "remap_DF = pd.DataFrame(remap_list)\n",
    "\n",
    "## Check\n",
    "## print_full(remap_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      position      score  seq_len                           species  \\\n",
      "34          10   5.454241      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "35         144  10.457056      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "36        -697   8.946094      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "37        -625   5.243600      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "38        -536   6.417715      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "39         336   3.702168      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "40         377   3.491528      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "41        -469   3.414712      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "42        -461   3.296591      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "43         423   3.794091      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "44        -416   9.909568      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "45         455   4.094957      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "46        -404   5.787577      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "47         521   3.793224      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "48        -301   9.909568      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "49         570   3.509995      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "50         599   3.794091      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "51         639   7.881554      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "52         669   3.412068      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "53         687   3.285077      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "54        -166   4.473469      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "55        -116   5.920559      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "56         -58   6.380240      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "57         814   5.505177      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "58         815   3.285077      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "59         856   3.875044      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "87          10   5.013668      862  ludwig_eve-striped-2||MEMB002C|+   \n",
      "88         150  10.457056      862  ludwig_eve-striped-2||MEMB002C|+   \n",
      "89        -678   8.946094      862  ludwig_eve-striped-2||MEMB002C|+   \n",
      "90        -606   5.243600      862  ludwig_eve-striped-2||MEMB002C|+   \n",
      "...        ...        ...      ...                               ...   \n",
      "8231       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8232       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8233       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8234       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8235       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8236       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8237       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8238       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8239       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8240       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8241       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8242       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8243       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8244       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8245       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8246       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8247       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8248       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8249       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8250       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8251       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8252       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8253       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8254       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8255       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8256       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8257       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8258       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8259       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "8260       NaN        NaN      NaN  ludwig_eve-striped-2||memb003f|+   \n",
      "\n",
      "      raw_position    strand  align_position  file  \n",
      "34              10  positive             NaN  file  \n",
      "35             144  positive             NaN  file  \n",
      "36             171  negative             NaN  file  \n",
      "37             243  negative             NaN  file  \n",
      "38             332  negative             NaN  file  \n",
      "39             336  positive             NaN  file  \n",
      "40             377  positive             NaN  file  \n",
      "41             399  negative             NaN  file  \n",
      "42             407  negative             NaN  file  \n",
      "43             423  positive             NaN  file  \n",
      "44             452  negative             NaN  file  \n",
      "45             455  positive             NaN  file  \n",
      "46             464  negative             NaN  file  \n",
      "47             521  positive             NaN  file  \n",
      "48             567  negative             NaN  file  \n",
      "49             570  positive             NaN  file  \n",
      "50             599  positive             NaN  file  \n",
      "51             639  positive             NaN  file  \n",
      "52             669  positive             NaN  file  \n",
      "53             687  positive             NaN  file  \n",
      "54             702  negative             NaN  file  \n",
      "55             752  negative             NaN  file  \n",
      "56             810  negative             NaN  file  \n",
      "57             814  positive             NaN  file  \n",
      "58             815  positive             NaN  file  \n",
      "59             856  positive             NaN  file  \n",
      "87              10  positive             NaN  file  \n",
      "88             150  positive             NaN  file  \n",
      "89             184  negative             NaN  file  \n",
      "90             256  negative             NaN  file  \n",
      "...            ...       ...             ...   ...  \n",
      "8231           838       NaN            1078  file  \n",
      "8232           839       NaN            1079  file  \n",
      "8233           840       NaN            1080  file  \n",
      "8234           841       NaN            1081  file  \n",
      "8235           842       NaN            1082  file  \n",
      "8236           843       NaN            1083  file  \n",
      "8237           844       NaN            1084  file  \n",
      "8238           845       NaN            1085  file  \n",
      "8239           846       NaN            1112  file  \n",
      "8240           847       NaN            1113  file  \n",
      "8241           848       NaN            1114  file  \n",
      "8242           849       NaN            1115  file  \n",
      "8243           850       NaN            1116  file  \n",
      "8244           851       NaN            1117  file  \n",
      "8245           852       NaN            1118  file  \n",
      "8246           853       NaN            1119  file  \n",
      "8247           854       NaN            1120  file  \n",
      "8248           855       NaN            1121  file  \n",
      "8249           856       NaN            1122  file  \n",
      "8250           857       NaN            1123  file  \n",
      "8251           858       NaN            1124  file  \n",
      "8252           859       NaN            1125  file  \n",
      "8253           860       NaN            1126  file  \n",
      "8254           861       NaN            1127  file  \n",
      "8255           862       NaN            1128  file  \n",
      "8256           863       NaN            1129  file  \n",
      "8257           864       NaN            1130  file  \n",
      "8258           865       NaN            1131  file  \n",
      "8259           866       NaN            1132  file  \n",
      "8260           867       NaN            1133  file  \n",
      "\n",
      "[8261 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "## Merge both datasets\n",
    "\n",
    "## Check first\n",
    "## print(position_DF.shape)\n",
    "## print(remap_DF.shape)\n",
    "\n",
    "## Merge - all sites\n",
    "TFBS_map_DF_all = pd.merge(position_DF, remap_DF, on=['species', 'raw_position'], how='outer')\n",
    "\n",
    "## Sort\n",
    "TFBS_map_DF_all = TFBS_map_DF_all.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_all)\n",
    "## print(TFBS_map_DF_all.shape)\n",
    "\n",
    "# Merge - only signal \n",
    "TFBS_map_DF_only_signal = pd.merge(position_DF, remap_DF, on=['species', 'raw_position'], how='inner')\n",
    "TFBS_map_DF_only_signal = TFBS_map_DF_only_signal.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "## To quickly check if species share similar TFBS positions\n",
    "## print_full(TFBS_map_DF_only_signal.sort_values(by=['raw_position'], ascending=[True]))\n",
    "\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_only_signal)\n",
    "## print(TFBS_map_DF_only_signal.shape)\n",
    "\n",
    "print(TFBS_map_DF_all)\n",
    "\n",
    "## Write out Files\n",
    "## TFBS_map_DF_all.to_csv('../data/outputs/TFBS_map_DF_all_bicoid_test.csv', sep='\\t', na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                               species  presence  align_position    strand\n",
      "34    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "35    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "36    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "37    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "38    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "39    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "40    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "41    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "42    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "43    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "44    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "45    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "46    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "47    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "48    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "49    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "50    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "51    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "52    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "53    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "54    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "55    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "56    ludwig_eve-striped-2||MEMB002A|+         1             NaN  negative\n",
      "57    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "58    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "59    ludwig_eve-striped-2||MEMB002A|+         1             NaN  positive\n",
      "87    ludwig_eve-striped-2||MEMB002C|+         1             NaN  positive\n",
      "88    ludwig_eve-striped-2||MEMB002C|+         1             NaN  positive\n",
      "89    ludwig_eve-striped-2||MEMB002C|+         1             NaN  negative\n",
      "90    ludwig_eve-striped-2||MEMB002C|+         1             NaN  negative\n",
      "...                                ...       ...             ...       ...\n",
      "8231  ludwig_eve-striped-2||memb003f|+         0            1078       NaN\n",
      "8232  ludwig_eve-striped-2||memb003f|+         0            1079       NaN\n",
      "8233  ludwig_eve-striped-2||memb003f|+         0            1080       NaN\n",
      "8234  ludwig_eve-striped-2||memb003f|+         0            1081       NaN\n",
      "8235  ludwig_eve-striped-2||memb003f|+         0            1082       NaN\n",
      "8236  ludwig_eve-striped-2||memb003f|+         0            1083       NaN\n",
      "8237  ludwig_eve-striped-2||memb003f|+         0            1084       NaN\n",
      "8238  ludwig_eve-striped-2||memb003f|+         0            1085       NaN\n",
      "8239  ludwig_eve-striped-2||memb003f|+         0            1112       NaN\n",
      "8240  ludwig_eve-striped-2||memb003f|+         0            1113       NaN\n",
      "8241  ludwig_eve-striped-2||memb003f|+         0            1114       NaN\n",
      "8242  ludwig_eve-striped-2||memb003f|+         0            1115       NaN\n",
      "8243  ludwig_eve-striped-2||memb003f|+         0            1116       NaN\n",
      "8244  ludwig_eve-striped-2||memb003f|+         0            1117       NaN\n",
      "8245  ludwig_eve-striped-2||memb003f|+         0            1118       NaN\n",
      "8246  ludwig_eve-striped-2||memb003f|+         0            1119       NaN\n",
      "8247  ludwig_eve-striped-2||memb003f|+         0            1120       NaN\n",
      "8248  ludwig_eve-striped-2||memb003f|+         0            1121       NaN\n",
      "8249  ludwig_eve-striped-2||memb003f|+         0            1122       NaN\n",
      "8250  ludwig_eve-striped-2||memb003f|+         0            1123       NaN\n",
      "8251  ludwig_eve-striped-2||memb003f|+         0            1124       NaN\n",
      "8252  ludwig_eve-striped-2||memb003f|+         0            1125       NaN\n",
      "8253  ludwig_eve-striped-2||memb003f|+         0            1126       NaN\n",
      "8254  ludwig_eve-striped-2||memb003f|+         0            1127       NaN\n",
      "8255  ludwig_eve-striped-2||memb003f|+         0            1128       NaN\n",
      "8256  ludwig_eve-striped-2||memb003f|+         0            1129       NaN\n",
      "8257  ludwig_eve-striped-2||memb003f|+         0            1130       NaN\n",
      "8258  ludwig_eve-striped-2||memb003f|+         0            1131       NaN\n",
      "8259  ludwig_eve-striped-2||memb003f|+         0            1132       NaN\n",
      "8260  ludwig_eve-striped-2||memb003f|+         0            1133       NaN\n",
      "\n",
      "[8261 rows x 4 columns]\n",
      "align_position                    6   7   8   9   10  11  12  13  14  15 ...  \\\n",
      "species                                                                  ...   \n",
      "ludwig_eve-striped-2||memb002a|+   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb002c|+   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb002d|+   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb002e|-   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb002f|+   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb003b|+   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb003c|-   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb003d|-   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||memb003f|+   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "\n",
      "align_position                    30  31  32  33  34  35  36  37  38  39  \n",
      "species                                                                   \n",
      "ludwig_eve-striped-2||memb002a|+   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb002c|+   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb002d|+   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb002e|-   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb002f|+   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb003b|+   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb003c|-   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb003d|-   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||memb003f|+   0   0   0   0   0   0   0   0   0   0  \n",
      "\n",
      "[9 rows x 34 columns]\n"
     ]
    }
   ],
   "source": [
    "###############################\n",
    "## Print binary info for each species\n",
    "################################\n",
    "\n",
    "# Create new column, 1 if TFBS is present, 0 if absent\n",
    "TFBS_map_DF_all['presence'] = 0 # For some reason you have to initaite the column first.\n",
    "TFBS_map_DF_all['presence'] = TFBS_map_DF_all.apply(lambda x: x.notnull(), axis=1)\n",
    "TFBS_map_DF_all['presence'] = TFBS_map_DF_all['presence'].astype(int)\n",
    "\n",
    "## Check   \n",
    "## print_full(TFBS_map_DF_all)\n",
    "\n",
    "## Create new dataframe\n",
    "\n",
    "## Check First\n",
    "## list(TFBS_map_DF_all.columns)\n",
    "\n",
    "TFBS_map_DF_binary = TFBS_map_DF_all[['species', 'presence', 'align_position', 'strand']].copy()\n",
    "\n",
    "## Subset on strand\n",
    "\n",
    "TFBS_map_DF_binary_positive = TFBS_map_DF_binary['strand'] == \"positive\"\n",
    "TFBS_map_DF_binary_negative = TFBS_map_DF_binary['strand'] == \"negative\"\n",
    "\n",
    "## Check\n",
    "print(TFBS_map_DF_binary)\n",
    "\n",
    "## Now long to wide\n",
    "## [ ] NaNs are introduced here and not sure why and it is super annoying.\n",
    "## - Maybe it has something to do with the negative and positive strands. These should be subsetted first.\n",
    "## There should be two presense and absence...or maybe a 2 presented? This get back to the range issue. \n",
    "## We should have the 1s maybe represent TFBS range, not just starting position.\n",
    "\n",
    "TFBS_map_DF_binary = TFBS_map_DF_binary.pivot_table(index='species', columns='align_position', values='presence')\n",
    "\n",
    "print(TFBS_map_DF_binary.iloc[:,6:40])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "####################\n",
    "## Attach input files name as a column\n",
    "#####################\n",
    "\n",
    "## Ideally I would attach the file name of the 1. raw sequence and 2. the motif being tested\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
