{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bring in Alignment for mapping\n",
    "\n",
    "This program will map TFBS using the Biopython's motif package.\n",
    "\n",
    "**Inputs**: \n",
    "1. before alignment (fasta) \n",
    "2. after alignment (fasta) \n",
    "3. TFBS Position Frequency Matrix.\n",
    "\n",
    "**Outputs**:\n",
    "1. `.csv` file that outputs found TFBSs at each position, if any, in alignment (position, score, species, raw_position, strand motif_found)\n",
    "2. `.csv` file that outputs only TFBS found.\n",
    "\n",
    "**To Do**:\n",
    "\n",
    "- [x] **Fix bug where species are labelled wrong only in reverse strand (how is this possible)?**\n",
    "    - I arbitrarily set the length. An artifact when I was looking at one sequence at a time.\n",
    "- [ ] **Why are there NaNs when I pivot the table? Weird.**\n",
    "- [x] Make Vector that attaches\n",
    "- [ ] Attach real species name\n",
    "- [ ] Loop through all files in directory\n",
    "- [ ] Write file name in last column\n",
    "- [ ] Append to a file \n",
    "- [ ] Have all of the important checks read out into a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import motifs\n",
    "from Bio import SeqIO \n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import MutableSeq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC, generic_dna, generic_protein\n",
    "from collections import defaultdict\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "[Errno 2] No such file or directory: '-f'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-de0fd294a820>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m## read in alignment as a list of sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0malignment\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;31m#motif = motifs.read(open(sys.argv[2],\"pfm\")\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python2.7/site-packages/Bio/SeqIO/__init__.pyc\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    581\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid alphabet, %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    582\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 583\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mas_handle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    584\u001b[0m         \u001b[0;31m# Map the file format to a sequence iterator:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    585\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python2.7/contextlib.pyc\u001b[0m in \u001b[0;36m__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__enter__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"generator didn't yield\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda/lib/python2.7/site-packages/Bio/File.pyc\u001b[0m in \u001b[0;36mas_handle\u001b[0;34m(handleish, mode, **kwargs)\u001b[0m\n\u001b[1;32m     88\u001b[0m                 \u001b[0;32myield\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 90\u001b[0;31m             \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandleish\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     91\u001b[0m                 \u001b[0;32myield\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     92\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: [Errno 2] No such file or directory: '-f'"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## sys Inputs - to do\n",
    "#####################\n",
    "\n",
    "## So I can run in shell and loop through sequences AND motifs to get a giant dataset\n",
    "## Need to get the alignment to raw sequence\n",
    "\n",
    "## read in alignment as a list of sequences\n",
    "alignment = list(SeqIO.parse(sys.argv[1], \"fasta\"))\n",
    "#motif = motifs.read(open(sys.argv[2],\"pfm\")\n",
    "\n",
    "## alignment = list(SeqIO.parse(\"../data/fasta/output_ludwig_eve-striped-2.fa\", \"fasta\"))\n",
    "motif = motifs.read(open(\"../data/PWM/transpose_fm/bcd_FlyReg.fm\"),\"pfm\")\n",
    "\n",
    "raw_sequences = []\n",
    "for record in alignment:\n",
    "    raw_sequences.append(SeqRecord(record.seq.ungap(\"-\"), id = record.id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 9 records in alignment file\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n"
     ]
    }
   ],
   "source": [
    "##################\n",
    "## Input 1 - Alignment Input \n",
    "##################\n",
    "\n",
    "## Sort alphabetically\n",
    "alignment = [f for f in sorted(alignment, key=lambda x : x.id)]\n",
    "\n",
    "## Make all ids lowercase\n",
    "\n",
    "for record in alignment:\n",
    "    record.id = record.id.lower()\n",
    "\n",
    "## Check \n",
    "print(\"Found %i records in alignment file\" % len(alignment))\n",
    "\n",
    "## Sequence Length should be the same for all alignment sequences\n",
    "for seq in alignment:\n",
    "    print len(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 9 records in raw sequence file\n",
      "[Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCAAGAAGGATGCGTTGTTAC...CGC', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTGAACTAACTCGCAGGAGCAAGAAGGATGCGGATGCGT...CTG', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCAAGAAGGATGCGGATGCGT...CTG', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCAAGAAGGATGCGGATGCGT...CTG', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCGGCAAGAAGGATGCGTTTG...GGC', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCGGCAAGAAGGATGCGTTGT...GGC', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCGGCCAGAAGGATGCGTTGT...CGC', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTGAACTAACTCGCAGGAGCGGCAAGAAGGATGCGTTGT...CTA', IUPACUnambiguousDNA()), Seq('ATATAACCCAATAATTTTAACTAACTCGCAGGAGCGGCAAGAAGGATGCGTTGT...CTA', IUPACUnambiguousDNA())]\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## Input 2 - Raw Sequences Input\n",
    "#####################\n",
    "\n",
    "print(\"Found %i records in raw sequence file\" % len(raw_sequences))\n",
    "\n",
    "## make all IUPAC.IUPACUnambiguousDNA()\n",
    "raw_sequences_2 = []\n",
    "for seq in raw_sequences:\n",
    "    raw_sequences_2.append(Seq(str(seq.seq), IUPAC.IUPACUnambiguousDNA()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        0      1      2      3      4      5      6      7\n",
      "A:   0.19   0.17   0.88   0.92   0.04   0.04   0.06   0.12\n",
      "C:   0.37   0.08   0.04   0.02   0.02   0.87   0.52   0.25\n",
      "G:   0.08   0.04   0.04   0.04   0.33   0.02   0.08   0.37\n",
      "T:   0.37   0.71   0.04   0.02   0.62   0.08   0.35   0.27\n",
      "\n",
      "        0      1      2      3      4      5      6      7\n",
      "A:  -0.38  -0.53   1.82   1.88  -2.70  -2.70  -2.12  -1.12\n",
      "C:   0.55  -1.70  -2.70  -3.70  -3.70   1.79   1.05  -0.00\n",
      "G:  -1.70  -2.70  -2.70  -2.70   0.39  -3.70  -1.70   0.55\n",
      "T:   0.55   1.51  -2.70  -3.70   1.30  -1.70   0.47   0.11\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## Input 3 - Motif Input\n",
    "#####################\n",
    "## motif.weblogo(\"mymotif.png\")\n",
    "print(motif.counts)\n",
    "pwm = motif.counts.normalize(pseudocounts=0.0) # Doesn't change from pwm\n",
    "pssm = pwm.log_odds()\n",
    "print(pssm) # Why do I need log odds exactly?\n",
    "motif_length = len(motif) #for later retrival of nucleotide sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "######################\n",
    "## Searching for Motifs in Sequences\n",
    "######################\n",
    "\n",
    "## Returns a list of arrays with a score for each position\n",
    "## This give the score for each position\n",
    "## If you print the length you get the length of the sequence minus TFBS length. \n",
    "\n",
    "## Forward stand\n",
    "pssm_list = [ ]\n",
    "for seq in raw_sequences_2:\n",
    "    pssm_list.append(pssm.calculate(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patser Threshold is 3.262\n"
     ]
    }
   ],
   "source": [
    "########################## \n",
    "## Automatic Calculation of threshold\n",
    "##########################\n",
    "\n",
    "## Ideal to find something that automatically calculates, as\n",
    "## opposed to having human choosing.\n",
    "\n",
    "## Approximate calculation of appropriate thresholds for motif finding \n",
    "## Patser Threshold\n",
    "## It selects such a threshold that the log(fpr)=-ic(M) \n",
    "## note: the actual patser software uses natural logarithms instead of log_2, so the numbers \n",
    "## are not directly comparable. \n",
    "\n",
    "distribution = pssm.distribution(background=motif.background, precision=10**4)\n",
    "patser_threshold = distribution.threshold_patser() #for use later\n",
    "\n",
    "print(\"Patser Threshold is %5.3f\" % patser_threshold) # Calculates Paster threshold. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     position      score  seq_len                           species\n",
      "0          10   5.013668      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "1        -804   3.414712      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "2         145  10.457056      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "3        -740   8.946094      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "4        -664   3.572565      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "5        -598   3.594098      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "6        -580   5.417715      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "7         389   3.824103      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "8         420   4.672100      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "9        -479   4.374594      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "10        430   3.794091      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "11       -448   9.909568      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "12        460   4.094957      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "13       -436   5.787577      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "14        502   3.793224      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "15        523   3.774115      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "16       -357   9.909568      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "17        551   4.094957      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "18       -313   3.769742      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "19       -305   5.431520      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "20       -275   3.832565      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "21       -274   5.505177      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "22        635   5.016483      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "23        641  10.016483      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "24        649   3.357991      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "25       -220   3.824103      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "26       -138   5.920559      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "27        -96   3.995421      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "28        813   3.391992      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "29        -82   6.380240      905  ludwig_eve-striped-2||MEMB002F|+\n",
      "..        ...        ...      ...                               ...\n",
      "205       847   3.285077      898  ludwig_eve-striped-2||MEMB003D|-\n",
      "206       888   3.875044      898  ludwig_eve-striped-2||MEMB003D|-\n",
      "207        10   5.454241      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "208       174  10.457056      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "209      -719   8.946094      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "210      -652   5.243600      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "211      -585   3.594098      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "212      -567   6.417715      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "213       350   3.702168      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "214       444   3.794091      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "215      -439   9.909568      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "216       477   4.094957      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "217      -427   5.787577      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "218       518   5.020957      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "219      -376   4.736640      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "220      -329   9.909568      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "221       587   4.094957      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "222       597   3.346478      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "223      -294   4.374594      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "224      -262   4.389700      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "225       655   5.016483      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "226       661  10.016483      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "227      -241   8.959556      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "228       676   8.997031      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "229       760   3.559102      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "230      -115   5.920559      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "231       844   3.391992      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "232       -59   6.380240      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "233       858   5.612093      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "234       901   4.422532      913  ludwig_eve-striped-2||MEMB002D|+\n",
      "\n",
      "[235 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "###################################\n",
    "## Searching for motif in all raw_sequences\n",
    "#################################\n",
    "\n",
    "raw_id = []\n",
    "for seq in raw_sequences:\n",
    "    raw_id.append(seq.id)\n",
    "\n",
    "record_length = []\n",
    "for record in raw_sequences_2:\n",
    "    record_length.append(len(record))\n",
    "\n",
    "position_list = []\n",
    "for i in range(0,8):\n",
    "    for position, score in pssm.search(raw_sequences_2[i], threshold = patser_threshold):\n",
    "        positions = {'species': raw_id[i], 'score':score, 'position':position, 'seq_len': record_length[i] }\n",
    "        position_list.append(positions)\n",
    "        \n",
    "position_DF = pd.DataFrame(position_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     position      score  seq_len                           species  \\\n",
      "0          10   5.013668      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "1        -804   3.414712      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "2         145  10.457056      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "3        -740   8.946094      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "4        -664   3.572565      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "5        -598   3.594098      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "6        -580   5.417715      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "7         389   3.824103      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "8         420   4.672100      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "9        -479   4.374594      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "10        430   3.794091      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "11       -448   9.909568      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "12        460   4.094957      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "13       -436   5.787577      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "14        502   3.793224      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "15        523   3.774115      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "16       -357   9.909568      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "17        551   4.094957      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "18       -313   3.769742      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "19       -305   5.431520      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "20       -275   3.832565      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "21       -274   5.505177      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "22        635   5.016483      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "23        641  10.016483      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "24        649   3.357991      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "25       -220   3.824103      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "26       -138   5.920559      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "27        -96   3.995421      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "28        813   3.391992      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "29        -82   6.380240      905  ludwig_eve-striped-2||MEMB002F|+   \n",
      "..        ...        ...      ...                               ...   \n",
      "205       847   3.285077      898  ludwig_eve-striped-2||MEMB003D|-   \n",
      "206       888   3.875044      898  ludwig_eve-striped-2||MEMB003D|-   \n",
      "207        10   5.454241      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "208       174  10.457056      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "209      -719   8.946094      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "210      -652   5.243600      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "211      -585   3.594098      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "212      -567   6.417715      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "213       350   3.702168      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "214       444   3.794091      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "215      -439   9.909568      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "216       477   4.094957      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "217      -427   5.787577      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "218       518   5.020957      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "219      -376   4.736640      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "220      -329   9.909568      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "221       587   4.094957      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "222       597   3.346478      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "223      -294   4.374594      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "224      -262   4.389700      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "225       655   5.016483      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "226       661  10.016483      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "227      -241   8.959556      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "228       676   8.997031      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "229       760   3.559102      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "230      -115   5.920559      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "231       844   3.391992      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "232       -59   6.380240      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "233       858   5.612093      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "234       901   4.422532      913  ludwig_eve-striped-2||MEMB002D|+   \n",
      "\n",
      "     raw_position    strand  \n",
      "0              10  positive  \n",
      "1             101  negative  \n",
      "2             145  positive  \n",
      "3             165  negative  \n",
      "4             241  negative  \n",
      "5             307  negative  \n",
      "6             325  negative  \n",
      "7             389  positive  \n",
      "8             420  positive  \n",
      "9             426  negative  \n",
      "10            430  positive  \n",
      "11            457  negative  \n",
      "12            460  positive  \n",
      "13            469  negative  \n",
      "14            502  positive  \n",
      "15            523  positive  \n",
      "16            548  negative  \n",
      "17            551  positive  \n",
      "18            592  negative  \n",
      "19            600  negative  \n",
      "20            630  negative  \n",
      "21            631  negative  \n",
      "22            635  positive  \n",
      "23            641  positive  \n",
      "24            649  positive  \n",
      "25            685  negative  \n",
      "26            767  negative  \n",
      "27            809  negative  \n",
      "28            813  positive  \n",
      "29            823  negative  \n",
      "..            ...       ...  \n",
      "205           847  positive  \n",
      "206           888  positive  \n",
      "207            10  positive  \n",
      "208           174  positive  \n",
      "209           194  negative  \n",
      "210           261  negative  \n",
      "211           328  negative  \n",
      "212           346  negative  \n",
      "213           350  positive  \n",
      "214           444  positive  \n",
      "215           474  negative  \n",
      "216           477  positive  \n",
      "217           486  negative  \n",
      "218           518  positive  \n",
      "219           537  negative  \n",
      "220           584  negative  \n",
      "221           587  positive  \n",
      "222           597  positive  \n",
      "223           619  negative  \n",
      "224           651  negative  \n",
      "225           655  positive  \n",
      "226           661  positive  \n",
      "227           672  negative  \n",
      "228           676  positive  \n",
      "229           760  positive  \n",
      "230           798  negative  \n",
      "231           844  positive  \n",
      "232           854  negative  \n",
      "233           858  positive  \n",
      "234           901  positive  \n",
      "\n",
      "[235 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "#############################\n",
    "## Add strand and pos position information as columns to position_DF\n",
    "#############################\n",
    "position_list_pos = []\n",
    "for i, x in enumerate(position_DF['position']):\n",
    "    if x < 0:\n",
    "       position_list_pos.append(position_DF.loc[position_DF.index[i], 'seq_len'] + x)\n",
    "    else:\n",
    "       position_list_pos.append(x)\n",
    "\n",
    "## append to position_DF\n",
    "position_DF['raw_position'] = position_list_pos\n",
    "\n",
    "## print position_DF['raw_position']\n",
    "    \n",
    "## strand Column\n",
    "strand = []\n",
    "for x in position_DF['position']:\n",
    "    if x < 0:\n",
    "       strand.append(\"negative\")\n",
    "    else:\n",
    "       strand.append(\"positive\")\n",
    "    \n",
    "## append to position_DF\n",
    "position_DF['strand'] = strand\n",
    "\n",
    "## motif_found column\n",
    "## First turn into a list of strings\n",
    "raw_sequences_2_list = []\n",
    "for seq in raw_sequences_2:\n",
    "    raw_sequences_2_list.append(str(seq))\n",
    "\n",
    "## Now get all motifs found in sequences\n",
    "# motif_found = []\n",
    "# for x in position_DF['position']:\n",
    "    # motif_found.append(raw_sequences_2_list[i][x:x + motif_length])\n",
    "\n",
    "## append to position_DF\n",
    "# position_DF['motif_found'] = motif_found\n",
    "\n",
    "print position_DF\n",
    "\n",
    "## Check\n",
    "## len(motif_found)    \n",
    "## print(motif_found) \n",
    "## print(position_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##################\n",
    "## get alignment position \n",
    "#################\n",
    "\n",
    "## Need to map to the sequence alignment position\n",
    "\n",
    "remap_list = []\n",
    "nuc_list = ['A', 'a', 'G', 'g', 'C', 'c', 'T', 't', 'N', 'n']\n",
    "\n",
    "\n",
    "positions = {'score':score, 'position':position, 'species': i}\n",
    "position_list.append(positions)\n",
    "\n",
    "for i in range(0,9):\n",
    "    counter = 0\n",
    "    for xInd, x in enumerate(alignment[i].seq):    \n",
    "        if x in nuc_list:\n",
    "            remaps = {'raw_position': counter, 'align_position':xInd, 'species':alignment[i].id}\n",
    "            counter += 1\n",
    "            remap_list.append(remaps)\n",
    "            \n",
    "remap_DF = pd.DataFrame(remap_list)\n",
    "\n",
    "## Check\n",
    "## print_full(remap_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      position     score  seq_len                           species  \\\n",
      "1106       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1107       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1108       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1109       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1110       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1111       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1112       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1113       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1114       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1115       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "34          10  5.454241      868  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1116       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1117       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1118       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1119       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1120       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1121       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1122       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1123       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1124       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1125       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1126       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1127       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1128       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1129       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1130       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1131       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1132       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1133       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "1134       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB002A|+   \n",
      "...        ...       ...      ...                               ...   \n",
      "5315       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5316       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5317       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5318       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5319       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5320       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5321       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5322       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5323       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5324       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5325       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5326       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5327       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5328       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5329       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5330       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5331       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5332       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5333       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5334       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "177        858  4.422532      868  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5335       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5336       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5337       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5338       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5339       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5340       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5341       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5342       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "5343       NaN       NaN      NaN  ludwig_eve-striped-2||MEMB003F|+   \n",
      "\n",
      "      raw_position    strand  align_position  \n",
      "1106             0       NaN               0  \n",
      "1107             1       NaN               1  \n",
      "1108             2       NaN               2  \n",
      "1109             3       NaN               3  \n",
      "1110             4       NaN               4  \n",
      "1111             5       NaN               5  \n",
      "1112             6       NaN               6  \n",
      "1113             7       NaN               7  \n",
      "1114             8       NaN               8  \n",
      "1115             9       NaN               9  \n",
      "34              10  positive              10  \n",
      "1116            11       NaN              11  \n",
      "1117            12       NaN              12  \n",
      "1118            13       NaN              13  \n",
      "1119            14       NaN              14  \n",
      "1120            15       NaN              15  \n",
      "1121            16       NaN              16  \n",
      "1122            17       NaN              17  \n",
      "1123            18       NaN              18  \n",
      "1124            19       NaN              19  \n",
      "1125            20       NaN              20  \n",
      "1126            21       NaN              21  \n",
      "1127            22       NaN              22  \n",
      "1128            23       NaN              23  \n",
      "1129            24       NaN              24  \n",
      "1130            25       NaN              25  \n",
      "1131            26       NaN              26  \n",
      "1132            27       NaN              27  \n",
      "1133            28       NaN              28  \n",
      "1134            29       NaN              29  \n",
      "...            ...       ...             ...  \n",
      "5315           838       NaN            1078  \n",
      "5316           839       NaN            1079  \n",
      "5317           840       NaN            1080  \n",
      "5318           841       NaN            1081  \n",
      "5319           842       NaN            1082  \n",
      "5320           843       NaN            1083  \n",
      "5321           844       NaN            1084  \n",
      "5322           845       NaN            1085  \n",
      "5323           846       NaN            1112  \n",
      "5324           847       NaN            1113  \n",
      "5325           848       NaN            1114  \n",
      "5326           849       NaN            1115  \n",
      "5327           850       NaN            1116  \n",
      "5328           851       NaN            1117  \n",
      "5329           852       NaN            1118  \n",
      "5330           853       NaN            1119  \n",
      "5331           854       NaN            1120  \n",
      "5332           855       NaN            1121  \n",
      "5333           856       NaN            1122  \n",
      "5334           857       NaN            1123  \n",
      "177            858  positive            1124  \n",
      "5335           859       NaN            1125  \n",
      "5336           860       NaN            1126  \n",
      "5337           861       NaN            1127  \n",
      "5338           862       NaN            1128  \n",
      "5339           863       NaN            1129  \n",
      "5340           864       NaN            1130  \n",
      "5341           865       NaN            1131  \n",
      "5342           866       NaN            1132  \n",
      "5343           867       NaN            1133  \n",
      "\n",
      "[8026 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "## Merge both datasets\n",
    "\n",
    "## Check first\n",
    "## print(position_DF.shape)\n",
    "## print(remap_DF.shape)\n",
    "\n",
    "## Merge - all sites\n",
    "TFBS_map_DF_all = pd.merge(position_DF, remap_DF, on=['species', 'raw_position'], how='outer')\n",
    "\n",
    "## Sort\n",
    "TFBS_map_DF_all = TFBS_map_DF_all.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_all)\n",
    "## print(TFBS_map_DF_all.shape)\n",
    "\n",
    "# Merge - only signal \n",
    "TFBS_map_DF_only_signal = pd.merge(position_DF, remap_DF, on=['species', 'raw_position'], how='inner')\n",
    "TFBS_map_DF_only_signal = TFBS_map_DF_only_signal.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "## To quickly check if species share similar TFBS positions\n",
    "## print_full(TFBS_map_DF_only_signal.sort_values(by=['raw_position'], ascending=[True]))\n",
    "\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_only_signal)\n",
    "## print(TFBS_map_DF_only_signal.shape)\n",
    "\n",
    "\n",
    "print(TFBS_map_DF_all)\n",
    "\n",
    "## Write out Files\n",
    "## TFBS_map_DF_all.to_csv('../data/outputs/TFBS_map_DF_all_bicoid_test.csv', sep='\\t', na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                               species  presence  align_position    strand\n",
      "1106  ludwig_eve-striped-2||MEMB002A|+         0               0       NaN\n",
      "1107  ludwig_eve-striped-2||MEMB002A|+         0               1       NaN\n",
      "1108  ludwig_eve-striped-2||MEMB002A|+         0               2       NaN\n",
      "1109  ludwig_eve-striped-2||MEMB002A|+         0               3       NaN\n",
      "1110  ludwig_eve-striped-2||MEMB002A|+         0               4       NaN\n",
      "1111  ludwig_eve-striped-2||MEMB002A|+         0               5       NaN\n",
      "1112  ludwig_eve-striped-2||MEMB002A|+         0               6       NaN\n",
      "1113  ludwig_eve-striped-2||MEMB002A|+         0               7       NaN\n",
      "1114  ludwig_eve-striped-2||MEMB002A|+         0               8       NaN\n",
      "1115  ludwig_eve-striped-2||MEMB002A|+         0               9       NaN\n",
      "34    ludwig_eve-striped-2||MEMB002A|+         1              10  positive\n",
      "1116  ludwig_eve-striped-2||MEMB002A|+         0              11       NaN\n",
      "1117  ludwig_eve-striped-2||MEMB002A|+         0              12       NaN\n",
      "1118  ludwig_eve-striped-2||MEMB002A|+         0              13       NaN\n",
      "1119  ludwig_eve-striped-2||MEMB002A|+         0              14       NaN\n",
      "1120  ludwig_eve-striped-2||MEMB002A|+         0              15       NaN\n",
      "1121  ludwig_eve-striped-2||MEMB002A|+         0              16       NaN\n",
      "1122  ludwig_eve-striped-2||MEMB002A|+         0              17       NaN\n",
      "1123  ludwig_eve-striped-2||MEMB002A|+         0              18       NaN\n",
      "1124  ludwig_eve-striped-2||MEMB002A|+         0              19       NaN\n",
      "1125  ludwig_eve-striped-2||MEMB002A|+         0              20       NaN\n",
      "1126  ludwig_eve-striped-2||MEMB002A|+         0              21       NaN\n",
      "1127  ludwig_eve-striped-2||MEMB002A|+         0              22       NaN\n",
      "1128  ludwig_eve-striped-2||MEMB002A|+         0              23       NaN\n",
      "1129  ludwig_eve-striped-2||MEMB002A|+         0              24       NaN\n",
      "1130  ludwig_eve-striped-2||MEMB002A|+         0              25       NaN\n",
      "1131  ludwig_eve-striped-2||MEMB002A|+         0              26       NaN\n",
      "1132  ludwig_eve-striped-2||MEMB002A|+         0              27       NaN\n",
      "1133  ludwig_eve-striped-2||MEMB002A|+         0              28       NaN\n",
      "1134  ludwig_eve-striped-2||MEMB002A|+         0              29       NaN\n",
      "...                                ...       ...             ...       ...\n",
      "5315  ludwig_eve-striped-2||MEMB003F|+         0            1078       NaN\n",
      "5316  ludwig_eve-striped-2||MEMB003F|+         0            1079       NaN\n",
      "5317  ludwig_eve-striped-2||MEMB003F|+         0            1080       NaN\n",
      "5318  ludwig_eve-striped-2||MEMB003F|+         0            1081       NaN\n",
      "5319  ludwig_eve-striped-2||MEMB003F|+         0            1082       NaN\n",
      "5320  ludwig_eve-striped-2||MEMB003F|+         0            1083       NaN\n",
      "5321  ludwig_eve-striped-2||MEMB003F|+         0            1084       NaN\n",
      "5322  ludwig_eve-striped-2||MEMB003F|+         0            1085       NaN\n",
      "5323  ludwig_eve-striped-2||MEMB003F|+         0            1112       NaN\n",
      "5324  ludwig_eve-striped-2||MEMB003F|+         0            1113       NaN\n",
      "5325  ludwig_eve-striped-2||MEMB003F|+         0            1114       NaN\n",
      "5326  ludwig_eve-striped-2||MEMB003F|+         0            1115       NaN\n",
      "5327  ludwig_eve-striped-2||MEMB003F|+         0            1116       NaN\n",
      "5328  ludwig_eve-striped-2||MEMB003F|+         0            1117       NaN\n",
      "5329  ludwig_eve-striped-2||MEMB003F|+         0            1118       NaN\n",
      "5330  ludwig_eve-striped-2||MEMB003F|+         0            1119       NaN\n",
      "5331  ludwig_eve-striped-2||MEMB003F|+         0            1120       NaN\n",
      "5332  ludwig_eve-striped-2||MEMB003F|+         0            1121       NaN\n",
      "5333  ludwig_eve-striped-2||MEMB003F|+         0            1122       NaN\n",
      "5334  ludwig_eve-striped-2||MEMB003F|+         0            1123       NaN\n",
      "177   ludwig_eve-striped-2||MEMB003F|+         1            1124  positive\n",
      "5335  ludwig_eve-striped-2||MEMB003F|+         0            1125       NaN\n",
      "5336  ludwig_eve-striped-2||MEMB003F|+         0            1126       NaN\n",
      "5337  ludwig_eve-striped-2||MEMB003F|+         0            1127       NaN\n",
      "5338  ludwig_eve-striped-2||MEMB003F|+         0            1128       NaN\n",
      "5339  ludwig_eve-striped-2||MEMB003F|+         0            1129       NaN\n",
      "5340  ludwig_eve-striped-2||MEMB003F|+         0            1130       NaN\n",
      "5341  ludwig_eve-striped-2||MEMB003F|+         0            1131       NaN\n",
      "5342  ludwig_eve-striped-2||MEMB003F|+         0            1132       NaN\n",
      "5343  ludwig_eve-striped-2||MEMB003F|+         0            1133       NaN\n",
      "\n",
      "[8026 rows x 4 columns]\n",
      "align_position                    6   7   8   9   10  11  12  13  14  15 ...  \\\n",
      "species                                                                  ...   \n",
      "ludwig_eve-striped-2||MEMB002A|+   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB002C|+   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB002D|+   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB002E|-   0   0   0   0   0   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB002F|+   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB003B|+   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB003C|-   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB003D|-   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "ludwig_eve-striped-2||MEMB003F|+   0   0   0   0   1   0   0   0   0   0 ...   \n",
      "\n",
      "align_position                    30  31  32  33  34  35  36  37  38  39  \n",
      "species                                                                   \n",
      "ludwig_eve-striped-2||MEMB002A|+   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB002C|+   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB002D|+   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB002E|-   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB002F|+   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB003B|+   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB003C|-   0   0   0 NaN NaN NaN   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB003D|-   0   0   0   0   0   0   0   0   0   0  \n",
      "ludwig_eve-striped-2||MEMB003F|+   0   0   0   0   0   0   0   0   0   0  \n",
      "\n",
      "[9 rows x 34 columns]\n"
     ]
    }
   ],
   "source": [
    "###############################\n",
    "## Print binary info for each species\n",
    "################################\n",
    "\n",
    "# Create new column, 1 if TFBS is present, 0 if absent\n",
    "TFBS_map_DF_all['presence'] = 0 # For some reason you have to initaite the column first.\n",
    "TFBS_map_DF_all['presence'] = TFBS_map_DF_all.apply(lambda x: x.notnull(), axis=1)\n",
    "TFBS_map_DF_all['presence'] = TFBS_map_DF_all['presence'].astype(int)\n",
    "\n",
    "## Check   \n",
    "## print_full(TFBS_map_DF_all)\n",
    "\n",
    "## Create new dataframe\n",
    "\n",
    "## Check First\n",
    "## list(TFBS_map_DF_all.columns)\n",
    "\n",
    "TFBS_map_DF_binary = TFBS_map_DF_all[['species', 'presence', 'align_position', 'strand']].copy()\n",
    "\n",
    "## Subset on strand\n",
    "\n",
    "TFBS_map_DF_binary_positive = TFBS_map_DF_binary['strand'] == \"positive\"\n",
    "TFBS_map_DF_binary_negative = TFBS_map_DF_binary['strand'] == \"negative\"\n",
    "\n",
    "## Check\n",
    "print(TFBS_map_DF_binary)\n",
    "\n",
    "## Now long to wide\n",
    "## [ ] NaNs are introduced here and not sure why and it is super annoying.\n",
    "## - Maybe it has something to do with the negative and positive strands. These should be subsetted first.\n",
    "## There should be two presense and absence...or maybe a 2 presented? This get back to the range issue. \n",
    "## We should have the 1s maybe represent TFBS range, not just starting position.\n",
    "\n",
    "TFBS_map_DF_binary = TFBS_map_DF_binary.pivot_table(index='species', columns='align_position', values='presence')\n",
    "\n",
    "print(TFBS_map_DF_binary.iloc[:,6:40])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "####################\n",
    "## Attach input files name as a column\n",
    "#####################\n",
    "\n",
    "## Ideally I would attach the file name of the 1. raw sequence and 2. the motif being tested\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
