{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "#raw_to_alignment_key.py\n",
    "#Ciera Martinez\n",
    "\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, sys\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO \n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC, generic_dna, generic_protein\n",
    "from collections import defaultdict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ludwig_eve-striped-2||MEMB002F|+', 'ludwig_eve-striped-2||MEMB002A|+', 'ludwig_eve-striped-2||MEMB003C|-', 'ludwig_eve-striped-2||MEMB002C|+', 'ludwig_eve-striped-2||MEMB003B|+', 'ludwig_eve-striped-2||MEMB003F|+', 'ludwig_eve-striped-2||MEMB003D|-', 'ludwig_eve-striped-2||MEMB002D|+', 'ludwig_eve-striped-2||MEMB002E|-']\n",
      "[905, 868, 875, 862, 909, 868, 898, 913, 928]\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## sys Inputs - to do\n",
    "#####################\n",
    "\n",
    "## read in alignment and motif \n",
    "## alignment = list(SeqIO.parse(sys.argv[1], \"fasta\"))\n",
    "\n",
    "## Used later when marking output file\n",
    "## alignment_file_name =  os.path.basename(sys.argv[1])\n",
    "## print alignment_file_name\n",
    "\n",
    "## Manual For testing\n",
    "alignment = list(SeqIO.parse(\"../data/fasta/output_ludwig_eve-striped-2.fa\", \"fasta\"))\n",
    "\n",
    "raw_sequences = []\n",
    "for record in alignment:\n",
    "    raw_sequences.append(SeqRecord(record.seq.ungap(\"-\"), id = record.id))\n",
    "\n",
    "## make raw sequences all IUPAC.IUPACUnambiguousDNA()\n",
    "raw_sequences_2 = []\n",
    "for seq in raw_sequences:\n",
    "    raw_sequences_2.append(Seq(str(seq.seq), IUPAC.IUPACUnambiguousDNA()))\n",
    "\n",
    "###################################\n",
    "## Searching for motif in all raw_sequences\n",
    "#################################\n",
    "\n",
    "raw_id = []\n",
    "for seq in raw_sequences:\n",
    "    raw_id.append(seq.id)\n",
    "    \n",
    "record_length = []\n",
    "for record in raw_sequences_2:\n",
    "    record_length.append(len(record))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      align_position  raw_position                           species  \\\n",
      "0                  0             0  ludwig_eve-striped-2||MEMB002F|+   \n",
      "1                  1             1  ludwig_eve-striped-2||MEMB002F|+   \n",
      "2                  2             2  ludwig_eve-striped-2||MEMB002F|+   \n",
      "3                  3             3  ludwig_eve-striped-2||MEMB002F|+   \n",
      "4                  4             4  ludwig_eve-striped-2||MEMB002F|+   \n",
      "5                  5             5  ludwig_eve-striped-2||MEMB002F|+   \n",
      "6                  6             6  ludwig_eve-striped-2||MEMB002F|+   \n",
      "7                  7             7  ludwig_eve-striped-2||MEMB002F|+   \n",
      "8                  8             8  ludwig_eve-striped-2||MEMB002F|+   \n",
      "9                  9             9  ludwig_eve-striped-2||MEMB002F|+   \n",
      "10                10            10  ludwig_eve-striped-2||MEMB002F|+   \n",
      "11                11            11  ludwig_eve-striped-2||MEMB002F|+   \n",
      "12                12            12  ludwig_eve-striped-2||MEMB002F|+   \n",
      "13                13            13  ludwig_eve-striped-2||MEMB002F|+   \n",
      "14                14            14  ludwig_eve-striped-2||MEMB002F|+   \n",
      "15                15            15  ludwig_eve-striped-2||MEMB002F|+   \n",
      "16                16            16  ludwig_eve-striped-2||MEMB002F|+   \n",
      "17                17            17  ludwig_eve-striped-2||MEMB002F|+   \n",
      "18                18            18  ludwig_eve-striped-2||MEMB002F|+   \n",
      "19                19            19  ludwig_eve-striped-2||MEMB002F|+   \n",
      "20                20            20  ludwig_eve-striped-2||MEMB002F|+   \n",
      "21                21            21  ludwig_eve-striped-2||MEMB002F|+   \n",
      "22                22            22  ludwig_eve-striped-2||MEMB002F|+   \n",
      "23                23            23  ludwig_eve-striped-2||MEMB002F|+   \n",
      "24                24            24  ludwig_eve-striped-2||MEMB002F|+   \n",
      "25                25            25  ludwig_eve-striped-2||MEMB002F|+   \n",
      "26                26            26  ludwig_eve-striped-2||MEMB002F|+   \n",
      "27                27            27  ludwig_eve-striped-2||MEMB002F|+   \n",
      "28                28            28  ludwig_eve-striped-2||MEMB002F|+   \n",
      "29                29            29  ludwig_eve-striped-2||MEMB002F|+   \n",
      "...              ...           ...                               ...   \n",
      "7996            1080           898  ludwig_eve-striped-2||MEMB002E|-   \n",
      "7997            1081           899  ludwig_eve-striped-2||MEMB002E|-   \n",
      "7998            1082           900  ludwig_eve-striped-2||MEMB002E|-   \n",
      "7999            1083           901  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8000            1084           902  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8001            1085           903  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8002            1112           904  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8003            1113           905  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8004            1114           906  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8005            1115           907  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8006            1116           908  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8007            1117           909  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8008            1118           910  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8009            1119           911  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8010            1120           912  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8011            1121           913  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8012            1122           914  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8013            1123           915  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8014            1124           916  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8015            1125           917  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8016            1126           918  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8017            1127           919  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8018            1128           920  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8019            1129           921  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8020            1130           922  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8021            1131           923  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8022            1132           924  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8023            1133           925  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8024            1134           926  ludwig_eve-striped-2||MEMB002E|-   \n",
      "8025            1135           927  ludwig_eve-striped-2||MEMB002E|-   \n",
      "\n",
      "     alignment_file  \n",
      "0              test  \n",
      "1              test  \n",
      "2              test  \n",
      "3              test  \n",
      "4              test  \n",
      "5              test  \n",
      "6              test  \n",
      "7              test  \n",
      "8              test  \n",
      "9              test  \n",
      "10             test  \n",
      "11             test  \n",
      "12             test  \n",
      "13             test  \n",
      "14             test  \n",
      "15             test  \n",
      "16             test  \n",
      "17             test  \n",
      "18             test  \n",
      "19             test  \n",
      "20             test  \n",
      "21             test  \n",
      "22             test  \n",
      "23             test  \n",
      "24             test  \n",
      "25             test  \n",
      "26             test  \n",
      "27             test  \n",
      "28             test  \n",
      "29             test  \n",
      "...             ...  \n",
      "7996           test  \n",
      "7997           test  \n",
      "7998           test  \n",
      "7999           test  \n",
      "8000           test  \n",
      "8001           test  \n",
      "8002           test  \n",
      "8003           test  \n",
      "8004           test  \n",
      "8005           test  \n",
      "8006           test  \n",
      "8007           test  \n",
      "8008           test  \n",
      "8009           test  \n",
      "8010           test  \n",
      "8011           test  \n",
      "8012           test  \n",
      "8013           test  \n",
      "8014           test  \n",
      "8015           test  \n",
      "8016           test  \n",
      "8017           test  \n",
      "8018           test  \n",
      "8019           test  \n",
      "8020           test  \n",
      "8021           test  \n",
      "8022           test  \n",
      "8023           test  \n",
      "8024           test  \n",
      "8025           test  \n",
      "\n",
      "[8026 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "##################\n",
    "## get alignment position \n",
    "#################\n",
    "\n",
    "remap_list = []\n",
    "nuc_list = ['A', 'a', 'G', 'g', 'C', 'c', 'T', 't', 'N', 'n']\n",
    "\n",
    "for i in range(0,9):\n",
    "    counter = 0\n",
    "    for xInd, x in enumerate(alignment[i].seq):    \n",
    "        if x in nuc_list:\n",
    "            remaps = {'raw_position': counter, 'align_position':xInd, 'species':alignment[i].id}\n",
    "            counter += 1\n",
    "            remap_list.append(remaps)\n",
    "            \n",
    "remap_DF = pd.DataFrame(remap_list)\n",
    "\n",
    "\n",
    "## Attach File Name\n",
    "remap_DF['alignment_file'] = \"alignment_file_name\"\n",
    "\n",
    "## Write out Files\n",
    "TFBS_map_DF_all.to_csv('../data/outputs/raw_to_alignment_key_output' + alignment_file_name + \".csv\", sep='\\t', na_rep=\"NA\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
