{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bring in Alignment for mapping\n",
    "\n",
    "This program will map TFBS using the Biopython's motif package.\n",
    "\n",
    "**Inputs**: \n",
    "1. before alignment (fasta) \n",
    "2. after alignment (fasta) \n",
    "3. TFBS Position Frequency Matrix.\n",
    "\n",
    "**Outputs**:\n",
    "1. `.csv` file that outputs found TFBSs at each position, if any, in alignment (position, score, species, raw_position, strand motif_found)\n",
    "2. `.csv` file that outputs only TFBS found.\n",
    "\n",
    "**To Do**:\n",
    "\n",
    "- [ ] **Fix bug where species are labelled wrong only in reverse strand (how is this possible)?**\n",
    "- [ ] **Why are there NaNs when I pivot the table? Weird.**\n",
    "- [x] Make Vector that attaches\n",
    "- [ ] Attach real species name\n",
    "- [ ] Loop through all files in directory\n",
    "- [ ] Write file name in last column\n",
    "- [ ] Append to a file \n",
    "- [ ] Have all of the important checks read out into a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import motifs\n",
    "from Bio import SeqIO \n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC, generic_dna, generic_protein\n",
    "from collections import defaultdict\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "####################\n",
    "## Functions\n",
    "#####################\n",
    "\n",
    "## Print full DF\n",
    "def print_full(x):\n",
    "    pd.set_option('display.max_rows', len(x))\n",
    "    print(x)\n",
    "    pd.reset_option('display.max_rows')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#####################\n",
    "## sys Inputs - to do\n",
    "#####################\n",
    "\n",
    "## So I can run in shell and loop through sequences AND motifs to get a giant dataset\n",
    "\n",
    "### input1 = open(sys.argv[1]) #file that contains text\n",
    "## input2 = open(sys.argv[2]) #file that contains terms to search for"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 9 records in alignment file\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n",
      "1136\n"
     ]
    }
   ],
   "source": [
    "##################\n",
    "## Input 1 - Alignment Input \n",
    "##################\n",
    "\n",
    "## read in alignment as a list of sequences\n",
    "alignment = list(SeqIO.parse(\"../data/fasta/output_ludwig_eve-striped-2.fa\", \"fasta\"))\n",
    "\n",
    "## Sort alphabetically\n",
    "alignment = [f for f in sorted(alignment, key=lambda x : x.id)]\n",
    "\n",
    "## Make all ids lowercase\n",
    "\n",
    "for record in alignment:\n",
    "    record.id = record.id.lower()\n",
    "\n",
    "## Check \n",
    "print(\"Found %i records in alignment file\" % len(alignment))\n",
    "\n",
    "## Sequence Length should be the same for all alignment sequences\n",
    "for seq in alignment:\n",
    "    print len(seq)\n",
    "    \n",
    "    \n",
    "## Not being used.\n",
    "## Turn sequences into a list of strings\n",
    "## Note: They are no longer bio.seq.seq objects\n",
    "\n",
    "## alignment_string_list = []\n",
    "## for seq in alignment:\n",
    "    ##alignment_string_list.append(str(seq.seq))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 9 records in raw sequence file\n",
      "868\n",
      "862\n",
      "913\n",
      "928\n",
      "905\n",
      "909\n",
      "875\n",
      "898\n",
      "868\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## Input 2 - Raw Sequences Input\n",
    "#####################\n",
    "\n",
    "## This is used to search for TFBSs in sequences\n",
    "## [ ] Maybe I have to convert from raw??\n",
    "\n",
    "raw_sequences = list(SeqIO.parse(\"../data/fasta/ludwig_eve-striped-2.fasta\", \"fasta\"))\n",
    "print(\"Found %i records in raw sequence file\" % len(raw_sequences))\n",
    "\n",
    "## Sort alphabetically\n",
    "raw_sequences = [f for f in sorted(raw_sequences, key=lambda x : x.id)]\n",
    "\n",
    "raw_id = []\n",
    "for seq in raw_sequences:\n",
    "    raw_id.append(seq.id)\n",
    "    print len(seq)\n",
    "    \n",
    "## make all IUPAC.IUPACUnambiguousDNA()\n",
    "raw_sequences_2 = []\n",
    "for seq in raw_sequences:\n",
    "    raw_sequences_2.append(Seq(str(seq.seq), IUPAC.IUPACUnambiguousDNA()))\n",
    "\n",
    "#print(raw_sequences)\n",
    "#print(raw_sequences_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        0      1      2      3      4      5      6      7\n",
      "A:   0.19   0.17   0.88   0.92   0.04   0.04   0.06   0.12\n",
      "C:   0.37   0.08   0.04   0.02   0.02   0.87   0.52   0.25\n",
      "G:   0.08   0.04   0.04   0.04   0.33   0.02   0.08   0.37\n",
      "T:   0.37   0.71   0.04   0.02   0.62   0.08   0.35   0.27\n",
      "\n",
      "        0      1      2      3      4      5      6      7\n",
      "A:  -0.38  -0.53   1.82   1.88  -2.70  -2.70  -2.12  -1.12\n",
      "C:   0.55  -1.70  -2.70  -3.70  -3.70   1.79   1.05  -0.00\n",
      "G:  -1.70  -2.70  -2.70  -2.70   0.39  -3.70  -1.70   0.55\n",
      "T:   0.55   1.51  -2.70  -3.70   1.30  -1.70   0.47   0.11\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "## Input 3 - Motif Input\n",
    "#####################\n",
    "\n",
    "motif = motifs.read(open(\"../data/PWM/transpose_fm/bcd_FlyReg.fm\"),\"pfm\")\n",
    "## motif.weblogo(\"mymotif.png\")\n",
    "print(motif.counts)\n",
    "pwm = motif.counts.normalize(pseudocounts=0.0) # Doesn't change from pwm\n",
    "pssm = pwm.log_odds()\n",
    "print(pssm) # Why do I need log odds exactly?\n",
    "motif_length = len(motif) #for later retrival of nucleotide sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "######################\n",
    "## Searching for Motifs in Sequences\n",
    "######################\n",
    "\n",
    "## Returns a list of arrays with a score for each position\n",
    "\n",
    "## This give the score for each position\n",
    "## If you print the length you get the length of the sequence minus TFBS length. \n",
    "\n",
    "## Forward stand\n",
    "\n",
    "pssm_list = [ ]\n",
    "for seq in raw_sequences_2:\n",
    "    pssm_list.append(pssm.calculate(seq))\n",
    "\n",
    "## Check\n",
    "## You should notice that the length is total length of raw_sequence minus length of motif.\n",
    "## for seq in pssm_list:\n",
    "##    print len(seq)\n",
    "    \n",
    "## Reverse strand\n",
    "## rpssm = pssm.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patser Threshold is 3.262\n"
     ]
    }
   ],
   "source": [
    "########################## \n",
    "## Automatic Calculation of threshold\n",
    "##########################\n",
    "\n",
    "## Ideal to find something that automatically calculates, as\n",
    "## opposed to having human choosing.\n",
    "\n",
    "## Approximate calculation of appropriate thresholds for motif finding \n",
    "## Patser Threshold\n",
    "## It selects such a threshold that the log(fpr)=-ic(M) \n",
    "## note: the actual patser software uses natural logarithms instead of log_2, so the numbers \n",
    "## are not directly comparable. \n",
    "\n",
    "distribution = pssm.distribution(background=motif.background, precision=10**4)\n",
    "patser_threshold = distribution.threshold_patser() #for use later\n",
    "\n",
    "print(\"Patser Threshold is %5.3f\" % patser_threshold) # Calculates Paster threshold. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     position      score  seq_len                           species\n",
      "0          10   5.454241      868  ludwig_eve-striped-2||memb002a|+\n",
      "1         144  10.457056      868  ludwig_eve-striped-2||memb002a|+\n",
      "2        -697   8.946094      868  ludwig_eve-striped-2||memb002a|+\n",
      "3        -625   5.243600      868  ludwig_eve-striped-2||memb002a|+\n",
      "4        -536   6.417715      868  ludwig_eve-striped-2||memb002a|+\n",
      "5         336   3.702168      868  ludwig_eve-striped-2||memb002a|+\n",
      "6         377   3.491528      868  ludwig_eve-striped-2||memb002a|+\n",
      "7        -469   3.414712      868  ludwig_eve-striped-2||memb002a|+\n",
      "8        -461   3.296591      868  ludwig_eve-striped-2||memb002a|+\n",
      "9         423   3.794091      868  ludwig_eve-striped-2||memb002a|+\n",
      "10       -416   9.909568      868  ludwig_eve-striped-2||memb002a|+\n",
      "11        455   4.094957      868  ludwig_eve-striped-2||memb002a|+\n",
      "12       -404   5.787577      868  ludwig_eve-striped-2||memb002a|+\n",
      "13        521   3.793224      868  ludwig_eve-striped-2||memb002a|+\n",
      "14       -301   9.909568      868  ludwig_eve-striped-2||memb002a|+\n",
      "15        570   3.509995      868  ludwig_eve-striped-2||memb002a|+\n",
      "16        599   3.794091      868  ludwig_eve-striped-2||memb002a|+\n",
      "17        639   7.881554      868  ludwig_eve-striped-2||memb002a|+\n",
      "18        669   3.412068      868  ludwig_eve-striped-2||memb002a|+\n",
      "19        687   3.285077      868  ludwig_eve-striped-2||memb002a|+\n",
      "20       -166   4.473469      868  ludwig_eve-striped-2||memb002a|+\n",
      "21       -116   5.920559      868  ludwig_eve-striped-2||memb002a|+\n",
      "22        -58   6.380240      868  ludwig_eve-striped-2||memb002a|+\n",
      "23        814   5.505177      868  ludwig_eve-striped-2||memb002a|+\n",
      "24        815   3.285077      868  ludwig_eve-striped-2||memb002a|+\n",
      "25        856   3.875044      868  ludwig_eve-striped-2||memb002a|+\n",
      "26         10   5.013668      862  ludwig_eve-striped-2||memb002c|+\n",
      "27        150  10.457056      862  ludwig_eve-striped-2||memb002c|+\n",
      "28       -678   8.946094      862  ludwig_eve-striped-2||memb002c|+\n",
      "29       -606   5.243600      862  ludwig_eve-striped-2||memb002c|+\n",
      "..        ...        ...      ...                               ...\n",
      "204       863   3.875044      875  ludwig_eve-striped-2||memb003c|-\n",
      "205        10   5.013668      898  ludwig_eve-striped-2||memb003d|-\n",
      "206        76   4.559102      898  ludwig_eve-striped-2||memb003d|-\n",
      "207       165  10.457056      898  ludwig_eve-striped-2||memb003d|-\n",
      "208      -713   8.946094      898  ludwig_eve-striped-2||memb003d|-\n",
      "209      -631   5.243600      898  ludwig_eve-striped-2||memb003d|-\n",
      "210      -597   3.285077      898  ludwig_eve-striped-2||memb003d|-\n",
      "211      -565   3.594098      898  ludwig_eve-striped-2||memb003d|-\n",
      "212      -547   6.417715      898  ludwig_eve-striped-2||memb003d|-\n",
      "213      -508   4.580384      898  ludwig_eve-striped-2||memb003d|-\n",
      "214       415   3.491528      898  ludwig_eve-striped-2||memb003d|-\n",
      "215       450   3.794091      898  ludwig_eve-striped-2||memb003d|-\n",
      "216      -421   9.909568      898  ludwig_eve-striped-2||memb003d|-\n",
      "217       480   4.094957      898  ludwig_eve-striped-2||memb003d|-\n",
      "218      -409   5.787577      898  ludwig_eve-striped-2||memb003d|-\n",
      "219      -358   4.736640      898  ludwig_eve-striped-2||memb003d|-\n",
      "220      -311   9.909568      898  ludwig_eve-striped-2||memb003d|-\n",
      "221       590   3.509995      898  ludwig_eve-striped-2||memb003d|-\n",
      "222       600   3.346478      898  ludwig_eve-striped-2||memb003d|-\n",
      "223      -271   6.920028      898  ludwig_eve-striped-2||memb003d|-\n",
      "224      -243   4.389700      898  ludwig_eve-striped-2||memb003d|-\n",
      "225       659   5.016483      898  ludwig_eve-striped-2||memb003d|-\n",
      "226       665   9.909568      898  ludwig_eve-striped-2||memb003d|-\n",
      "227      -224   4.435994      898  ludwig_eve-striped-2||memb003d|-\n",
      "228       678   5.787577      898  ludwig_eve-striped-2||memb003d|-\n",
      "229      -112   5.920559      898  ludwig_eve-striped-2||memb003d|-\n",
      "230       -56   6.380240      898  ludwig_eve-striped-2||memb003d|-\n",
      "231       846   5.505177      898  ludwig_eve-striped-2||memb003d|-\n",
      "232       847   3.285077      898  ludwig_eve-striped-2||memb003d|-\n",
      "233       888   3.875044      898  ludwig_eve-striped-2||memb003d|-\n",
      "\n",
      "[234 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "###################################\n",
    "## Searching for motif in all raw_sequences\n",
    "#################################\n",
    "\n",
    "record_length = []\n",
    "for record in raw_sequences_2_list:\n",
    "    record_length.append(len(record))\n",
    "\n",
    "position_list = []\n",
    "for i in range(0,8):\n",
    "    for position, score in pssm.search(raw_sequences_2[i], threshold = patser_threshold):\n",
    "        positions = {'species': raw_id[i], 'score':score, 'position':position, 'seq_len': record_length[i] }\n",
    "        position_list.append(positions)\n",
    "        \n",
    "position_DF = pd.DataFrame(position_list)\n",
    "\n",
    "## Check\n",
    "print position_DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     position      score  seq_len                           species  \\\n",
      "0          10   5.454241      868  ludwig_eve-striped-2||memb002a|+   \n",
      "1         144  10.457056      868  ludwig_eve-striped-2||memb002a|+   \n",
      "2        -697   8.946094      868  ludwig_eve-striped-2||memb002a|+   \n",
      "3        -625   5.243600      868  ludwig_eve-striped-2||memb002a|+   \n",
      "4        -536   6.417715      868  ludwig_eve-striped-2||memb002a|+   \n",
      "5         336   3.702168      868  ludwig_eve-striped-2||memb002a|+   \n",
      "6         377   3.491528      868  ludwig_eve-striped-2||memb002a|+   \n",
      "7        -469   3.414712      868  ludwig_eve-striped-2||memb002a|+   \n",
      "8        -461   3.296591      868  ludwig_eve-striped-2||memb002a|+   \n",
      "9         423   3.794091      868  ludwig_eve-striped-2||memb002a|+   \n",
      "10       -416   9.909568      868  ludwig_eve-striped-2||memb002a|+   \n",
      "11        455   4.094957      868  ludwig_eve-striped-2||memb002a|+   \n",
      "12       -404   5.787577      868  ludwig_eve-striped-2||memb002a|+   \n",
      "13        521   3.793224      868  ludwig_eve-striped-2||memb002a|+   \n",
      "14       -301   9.909568      868  ludwig_eve-striped-2||memb002a|+   \n",
      "15        570   3.509995      868  ludwig_eve-striped-2||memb002a|+   \n",
      "16        599   3.794091      868  ludwig_eve-striped-2||memb002a|+   \n",
      "17        639   7.881554      868  ludwig_eve-striped-2||memb002a|+   \n",
      "18        669   3.412068      868  ludwig_eve-striped-2||memb002a|+   \n",
      "19        687   3.285077      868  ludwig_eve-striped-2||memb002a|+   \n",
      "20       -166   4.473469      868  ludwig_eve-striped-2||memb002a|+   \n",
      "21       -116   5.920559      868  ludwig_eve-striped-2||memb002a|+   \n",
      "22        -58   6.380240      868  ludwig_eve-striped-2||memb002a|+   \n",
      "23        814   5.505177      868  ludwig_eve-striped-2||memb002a|+   \n",
      "24        815   3.285077      868  ludwig_eve-striped-2||memb002a|+   \n",
      "25        856   3.875044      868  ludwig_eve-striped-2||memb002a|+   \n",
      "26         10   5.013668      862  ludwig_eve-striped-2||memb002c|+   \n",
      "27        150  10.457056      862  ludwig_eve-striped-2||memb002c|+   \n",
      "28       -678   8.946094      862  ludwig_eve-striped-2||memb002c|+   \n",
      "29       -606   5.243600      862  ludwig_eve-striped-2||memb002c|+   \n",
      "..        ...        ...      ...                               ...   \n",
      "204       863   3.875044      875  ludwig_eve-striped-2||memb003c|-   \n",
      "205        10   5.013668      898  ludwig_eve-striped-2||memb003d|-   \n",
      "206        76   4.559102      898  ludwig_eve-striped-2||memb003d|-   \n",
      "207       165  10.457056      898  ludwig_eve-striped-2||memb003d|-   \n",
      "208      -713   8.946094      898  ludwig_eve-striped-2||memb003d|-   \n",
      "209      -631   5.243600      898  ludwig_eve-striped-2||memb003d|-   \n",
      "210      -597   3.285077      898  ludwig_eve-striped-2||memb003d|-   \n",
      "211      -565   3.594098      898  ludwig_eve-striped-2||memb003d|-   \n",
      "212      -547   6.417715      898  ludwig_eve-striped-2||memb003d|-   \n",
      "213      -508   4.580384      898  ludwig_eve-striped-2||memb003d|-   \n",
      "214       415   3.491528      898  ludwig_eve-striped-2||memb003d|-   \n",
      "215       450   3.794091      898  ludwig_eve-striped-2||memb003d|-   \n",
      "216      -421   9.909568      898  ludwig_eve-striped-2||memb003d|-   \n",
      "217       480   4.094957      898  ludwig_eve-striped-2||memb003d|-   \n",
      "218      -409   5.787577      898  ludwig_eve-striped-2||memb003d|-   \n",
      "219      -358   4.736640      898  ludwig_eve-striped-2||memb003d|-   \n",
      "220      -311   9.909568      898  ludwig_eve-striped-2||memb003d|-   \n",
      "221       590   3.509995      898  ludwig_eve-striped-2||memb003d|-   \n",
      "222       600   3.346478      898  ludwig_eve-striped-2||memb003d|-   \n",
      "223      -271   6.920028      898  ludwig_eve-striped-2||memb003d|-   \n",
      "224      -243   4.389700      898  ludwig_eve-striped-2||memb003d|-   \n",
      "225       659   5.016483      898  ludwig_eve-striped-2||memb003d|-   \n",
      "226       665   9.909568      898  ludwig_eve-striped-2||memb003d|-   \n",
      "227      -224   4.435994      898  ludwig_eve-striped-2||memb003d|-   \n",
      "228       678   5.787577      898  ludwig_eve-striped-2||memb003d|-   \n",
      "229      -112   5.920559      898  ludwig_eve-striped-2||memb003d|-   \n",
      "230       -56   6.380240      898  ludwig_eve-striped-2||memb003d|-   \n",
      "231       846   5.505177      898  ludwig_eve-striped-2||memb003d|-   \n",
      "232       847   3.285077      898  ludwig_eve-striped-2||memb003d|-   \n",
      "233       888   3.875044      898  ludwig_eve-striped-2||memb003d|-   \n",
      "\n",
      "     raw_position    strand motif_found  \n",
      "0              10  positive    ataatttt  \n",
      "1             144  positive    tgtgtgtg  \n",
      "2             171  negative    tttgggcc  \n",
      "3             243  negative    ttattgcc  \n",
      "4             332  negative    gaattccg  \n",
      "5             336  positive    ctttcaat  \n",
      "6             377  positive    ggagacag  \n",
      "7             399  negative    aaagtcag  \n",
      "8             407  negative    gagattcc  \n",
      "9             423  positive    ggttgcaa  \n",
      "10            452  negative    tagccgag  \n",
      "11            455  positive    ccagtttg  \n",
      "12            464  negative    tgacttgg  \n",
      "13            521  positive    tccagccg  \n",
      "14            567  negative    ggacatat  \n",
      "15            570  positive    ataataat  \n",
      "16            599  positive    acatatcc  \n",
      "17            639  positive    gttatttt  \n",
      "18            669  positive    tcccccag  \n",
      "19            687  positive    gcaggtag  \n",
      "20            702  negative    tccccccc  \n",
      "21            752  negative    gggcatga  \n",
      "22            810  negative    ttcaaatt  \n",
      "23            814  positive    gcccaacg  \n",
      "24            815  positive    cccaacga  \n",
      "25            856  positive    gcaactgg  \n",
      "26             10  positive    ataatttt  \n",
      "27            150  positive    tgtgtttg  \n",
      "28            184  negative    atttccct  \n",
      "29            256  negative    aatcgaag  \n",
      "..            ...       ...         ...  \n",
      "204           863  positive    gggccgtt  \n",
      "205            10  positive    ataatttt  \n",
      "206            76  positive    gcaatcct  \n",
      "207           165  positive    ttaatccg  \n",
      "208           185  negative    cagattat  \n",
      "209           267  negative    tagatttt  \n",
      "210           301  negative    gagttttg  \n",
      "211           333  negative    tggctttc  \n",
      "212           351  negative    ggaattaa  \n",
      "213           390  negative    aggctaaa  \n",
      "214           415  positive    aaaaaccg  \n",
      "215           450  positive    ctaaccca  \n",
      "216           477  negative    gggattag  \n",
      "217           480  positive    attagccg  \n",
      "218           489  negative    gggcttga  \n",
      "219           540  negative    ccgatttt  \n",
      "220           587  negative    gggattag  \n",
      "221           590  positive    attagctg  \n",
      "222           600  positive    catatccc  \n",
      "223           627  negative    cagctttg  \n",
      "224           655  negative    tgacttaa  \n",
      "225           659  positive    ttaaccct  \n",
      "226           665  positive    ctaatccc  \n",
      "227           674  negative    cagctcaa  \n",
      "228           678  positive    tcaagccc  \n",
      "229           786  negative    atgattat  \n",
      "230           842  negative    caaattaa  \n",
      "231           846  positive    ttaagtcc  \n",
      "232           847  positive    taagtccc  \n",
      "233           888  positive    attatctc  \n",
      "\n",
      "[234 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "#############################\n",
    "## Add strand and pos position information as columns to position_DF\n",
    "##############################\n",
    "\n",
    "## raw_position column\n",
    "## change position to positive\n",
    "## [ ] Fix length problem\n",
    "    \n",
    "position_list_pos = []\n",
    "\n",
    "for i, x in enumerate(position_DF['position']):\n",
    "    if x < 0:\n",
    "       position_list_pos.append(position_DF.loc[position_DF.index[i], 'seq_len'] + x)\n",
    "    else:\n",
    "       position_list_pos.append(x)\n",
    "\n",
    "## append to position_DF\n",
    "position_DF['raw_position'] = position_list_pos\n",
    "\n",
    "## print position_DF['raw_position']\n",
    "    \n",
    "## strand Column\n",
    "strand = []\n",
    "for x in position_DF['position']:\n",
    "    if x < 0:\n",
    "       strand.append(\"negative\")\n",
    "    else:\n",
    "       strand.append(\"positive\")\n",
    "    \n",
    "## append to position_DF\n",
    "position_DF['strand'] = strand\n",
    "\n",
    "## motif_found column\n",
    "## First turn into a list of strings\n",
    "raw_sequences_2_list = []\n",
    "for seq in raw_sequences_2:\n",
    "    raw_sequences_2_list.append(str(seq))\n",
    "\n",
    "## Now get all motifs found in sequences\n",
    "# motif_found = []\n",
    "# for x in position_DF['position']:\n",
    "    # motif_found.append(raw_sequences_2_list[i][x:x + motif_length])\n",
    "\n",
    "## append to position_DF\n",
    "# position_DF['motif_found'] = motif_found\n",
    "\n",
    "print position_DF\n",
    "\n",
    "## Check\n",
    "## len(motif_found)    \n",
    "## print(motif_found) \n",
    "## print(position_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##################\n",
    "## get alignment position \n",
    "#################\n",
    "\n",
    "## Need to map to the sequence alignment position\n",
    "\n",
    "remap_list = []\n",
    "nuc_list = ['A', 'a', 'G', 'g', 'C', 'c', 'T', 't', 'N', 'n']\n",
    "\n",
    "\n",
    "positions = {'score':score, 'position':position, 'species': i}\n",
    "position_list.append(positions)\n",
    "\n",
    "for i in range(0,9):\n",
    "    counter = 0\n",
    "    for xInd, x in enumerate(alignment[i].seq):    \n",
    "        if x in nuc_list:\n",
    "            remaps = {'raw_position': counter, 'align_position':xInd, 'species':alignment[i].id}\n",
    "            counter += 1\n",
    "            remap_list.append(remaps)\n",
    "            \n",
    "remap_DF = pd.DataFrame(remap_list)\n",
    "\n",
    "## Check\n",
    "## print_full(remap_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "TTATTGCC\n",
      "ttgaacaa\n",
      "ttattgcc\n",
      "agatttta\n",
      "cttgaaca\n",
      "aacaatcg\n",
      "ttgccgca\n",
      "ttattgcc\n",
      "cgagttag\n",
      "gtgctcta\n"
     ]
    }
   ],
   "source": [
    "## A way to check sequences\n",
    "## [ ] Write better test\n",
    "\n",
    "for seq in alignment:\n",
    "    print(seq.seq[345:353])\n",
    "    \n",
    "for seq in raw_sequences_2_list:\n",
    "    print(seq[262:270])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Merge both datasets\n",
    "\n",
    "## Check first\n",
    "## print(position_DF.shape)\n",
    "## print(remap_DF.shape)\n",
    "\n",
    "## Merge - all sites\n",
    "TFBS_map_DF_all = pd.merge(position_DF, remap_DF, on=['species', 'raw_position'], how='outer')\n",
    "\n",
    "## Sort\n",
    "TFBS_map_DF_all = TFBS_map_DF_all.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_all)\n",
    "## print(TFBS_map_DF_all.shape)\n",
    "\n",
    "# Merge - only signal \n",
    "TFBS_map_DF_only_signal = pd.merge(position_DF, remap_DF, on=['species', 'raw_position'], how='inner')\n",
    "TFBS_map_DF_only_signal = TFBS_map_DF_only_signal.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "## To quickly check if species share similar TFBS positions\n",
    "## print_full(TFBS_map_DF_only_signal.sort_values(by=['raw_position'], ascending=[True]))\n",
    "\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_only_signal)\n",
    "## print(TFBS_map_DF_only_signal.shape)\n",
    "\n",
    "\n",
    "print_full(TFBS_map_DF_all)\n",
    "\n",
    "## Write out Files\n",
    "TFBS_map_DF_all.to_csv('../data/outputs/TFBS_map_DF_all_bicoid_test.csv', sep='\\t', na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###############################\n",
    "## Print binary info for each species\n",
    "################################\n",
    "\n",
    "# Create new column, 1 if TFBS is present, 0 if absent\n",
    "TFBS_map_DF_all['presence'] = 0 # For some reason you have to initaite the column first.\n",
    "TFBS_map_DF_all['presence'] = TFBS_map_DF_all.apply(lambda x: x.notnull(), axis=1)\n",
    "TFBS_map_DF_all['presence'] = TFBS_map_DF_all['presence'].astype(int)\n",
    "\n",
    "## Check   \n",
    "## print_full(TFBS_map_DF_all)\n",
    "\n",
    "## Create new dataframe\n",
    "\n",
    "## Check First\n",
    "## list(TFBS_map_DF_all.columns)\n",
    "\n",
    "TFBS_map_DF_binary = TFBS_map_DF_all[['species', 'presence', 'align_position', 'strand']].copy()\n",
    "\n",
    "## Subset on strand\n",
    "\n",
    "TFBS_map_DF_binary_positive = TFBS_map_DF_binary['strand'] == \"positive\"\n",
    "TFBS_map_DF_binary_negative = TFBS_map_DF_binary['strand'] == \"negative\"\n",
    "\n",
    "## Check\n",
    "## print_full(TFBS_map_DF_binary)\n",
    "\n",
    "## Now long to wide\n",
    "## [ ] NaNs are introduced here and not sure why and It is super annoying.\n",
    "## - Maybe it has something to do with the negative and positive strands. These should be subsetted first. \n",
    "\n",
    "TFBS_map_DF_binary = TFBS_map_DF_binary.pivot_table(index='species', columns='align_position', values='presence')\n",
    "\n",
    "print_full(TFBS_map_DF_binary.iloc[:,6:15])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "####################\n",
    "## Attach input files name as a column\n",
    "#####################\n",
    "\n",
    "## Ideally I would attach the file name of the 1. raw sequence and 2. the motif being tested\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
